src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs

```diff
@@ -1220,36 +1313,35 @@ IList<ArraySegment<byte>> ConvertLockTokensToDeliveryTags(IEnumerable<Guid> lock
             return requestResponseAmqpLink;
         }
 
-        void CloseSession(ReceivingAmqpLink link)
+        void OnSessionReceiverLinkClosed(object sender, EventArgs e)
         {
-            link.Session.SafeClose();
+            ReceivingAmqpLink amqpLink = (ReceivingAmqpLink)sender;
+            if (amqpLink != null)
+            {
+                var exception = amqpLink.GetInnerException();
+                if (!(exception is SessionLockLostException))
+                {
+                    exception = new SessionLockLostException("Session lock lost. Accept a new session", exception);
+                }
+
+                this.LinkException = exception;
+                MessagingEventSource.Log.SessionReceiverLinkClosed(this.ClientId, this.SessionIdInternal, this.LinkException);
+            }
         }
 
-        void CloseRequestResponseSession(RequestResponseAmqpLink requestResponseAmqpLink)
+        IList<ArraySegment<byte>> ConvertLockTokensToDeliveryTags(IEnumerable<Guid> lockTokens)
         {
-            requestResponseAmqpLink.Session.SafeClose();
+            return lockTokens.Select(lockToken => new ArraySegment<byte>(lockToken.ToByteArray())).ToList();
```

Comment:
Returned value is a `List<ArraySegment<byte>>`, should the method signature be `List<...>` as well?